pycaw/tools/app_mixer/callbacks_and_manager.py

```diff
@@ -0,0 +1,112 @@
+    handles checking if session already exists
+    by using the session.InstanceIdentifier
+    (sometimes a session fires multiple times OnSessionCreated
+    even its already there)
+
+    The session.InstanceIdentifier is unique!
+    For every new session of an app unique!
+    For multiple sessions of an app unique!
+
+    it also handles:
+    - initiating new AppRow frames and passing there the newly created session.
+    - removing sessions from list
+    - unregister all (still active) sessions from callback
+    at 'shutdown' of this app
+
+    but mainly:
+
+    on_session_created() is fired by Windows everytime a new session registers
+
+    """
+
+    @classmethod
+    def __init__(cls, RootFrameInstance, AppRowClass):
+        cls.active = {}
+        cls.RootFrameInstance = RootFrameInstance
+        cls.AppRowClass = AppRowClass
+
+    @classmethod
+    def add_session(cls, session):
+        iid = session.InstanceIdentifier
+        if iid not in cls.active:
+            AppRowInstance = cls.AppRowClass(cls.RootFrameInstance,
+                                             session,
+                                             iid,
+                                             cls)
+
+            cls.active[iid] = AppRowInstance
+
+            # _______ ADD AUDIO SESSION CALLBACK _______
+            callback = AudioSessionCallback(AppRowInstance)
+            session.register_notification(callback)
+            # add callback to be able to listen to IAudioSessionEvents
+            # pass 'app_row' (new AppRow) to
+            # AudioSessionCallback -> AudioSessionEvents,
+            # so that the callback instance can call methods from this AppRow.
+
+    @classmethod
+    def remove_session(cls, iid):
+        if iid in cls.active:
+            del cls.active[iid]
+
+    @classmethod
+    def unregister_all(cls):
+        # since cls.active can always change (multithreading)
+        # instead of a for loop i present you the while loop:
+        # while cls.active contains any items,
+        # they get popped and unregistered
+        while cls.active:
+            # ________ REMOVES 1 ITEM FROM LIST ________
+            _, app_row = cls.active.popitem()
+
+            try:
+                app_row.audio_session.unregister_notification()
+            except TypeError as e:
+                # should raise an error if a session is closed
+                # between popitem() and unregister_notification()
+                # but if you are that lucky - you should go to the lottery
+                print("you won the lottery")
+                print(str(e))
+            print(":: :: :: unregister Session")
+
+    @classmethod
+    def on_session_created(cls, new_session):
+        print(':: OnSessionCreated\n'
+              f"new_session: {str(new_session)}")
+        cls.add_session(new_session)
+
+
+class AudioSessionCallback(AudioSessionEvents):
+    """Audio session callback"""
+    def __init__(self, AppRowInstance):
```

Comment:
If `AppRowInstance` is an instance it should be snake case

Reply:
okay ^^